cockatrice/src/filter_string.cpp

```diff
@@ -0,0 +1,217 @@
+
+FlexStringValue <- CompactStringSet / String / [(] StringList [)]
+CompactStringSet <- StringListString ([,+] StringListString)+
+
+NumericExpression <- NumericOperator ws? NumericValue
+NumericOperator <- [=:] / <[><!][=]?>
+NumericValue <- [0-9]+ 
+
+    )");
+
+std::once_flag init;
+
+void tracer(const char *name, const char *s, size_t n, const peg::SemanticValues, const peg::Context, const peg::any)
+{
+    std::cerr << "T " << name << "/" << s << "/" << n << std::endl;
+}
+
+FilterString::FilterString(const QString &expr)
+{
+    QByteArray ba = expr.toLocal8Bit();
+
+    std::call_once(init, []() {
+        search.log = [](size_t line, size_t col, const std::string &msg) {
+            std::cerr << line << ":" << col << ": " << msg << "\n";
+        };
+
+        auto pass = [](const peg::SemanticValues &sv) -> Filter {
+            return sv.size() > 0 ? sv[0].get<Filter>() : nullptr;
+        };
+
+        search["Start"] = pass;
+        search["QueryPartList"] = [](const peg::SemanticValues &sv) -> Filter {
+            return [=](CardData x) {
+                for (int i = 0; i < sv.size(); ++i) {
+                    if (!sv[i].get<Filter>()(x))
+                        return false;
+                }
+                return true;
+            };
+        };
+        search["ComplexQueryPart"] = [](const peg::SemanticValues &sv) -> Filter {
+            return [=](CardData x) {
+                for (int i = 0; i < sv.size(); ++i) {
+                    if (sv[i].get<Filter>()(x))
+                        return true;
+                }
+                return false;
+            };
+        };
+        search["SomewhatComplexQueryPart"] = pass;
+        search["QueryPart"] = pass;
+        search["NotQuery"] = [](const peg::SemanticValues &sv) -> Filter {
+            Filter dependent = sv[0].get<Filter>();
+            return [=](CardData x) -> bool { return !dependent(x); };
+        };
+        search["TypeQuery"] = [](const peg::SemanticValues &sv) -> Filter {
+            StringMatcher matcher = sv[0].get<StringMatcher>();
+            return [=](CardData x) -> bool { return matcher(x->getCardType()); };
+        };
+        search["StringValue"] = [](const peg::SemanticValues &sv) -> StringMatcher {
+            auto target = sv[0].get<QString>();
+            return [=](const QString &s) { return s.split(" ").contains(target, Qt::CaseInsensitive); };
+        };
+
+        search["String"] = [](const peg::SemanticValues &sv) -> QString {
+            if (sv.choice() == 0) {
+                return QString::fromStdString(sv.str());
+            } else {
+                return QString::fromStdString(sv.token(0));
+            }
+        };
+
+        search["NumericExpression"] = [](const peg::SemanticValues &sv) -> NumberMatcher {
+            auto arg = sv[1].get<int>();
+            auto op = sv[0].get<QString>();
+            return [=](int s) {
+                if (op == ">")
+                    return s > arg;
```

Comment:
I remember in the past for a school problem I was able to convert the op string to a real op without switches... maybe I'm mistaken?